### PR TITLE
fix crash in test_funcattrs

### PR DESCRIPTION
**Pull request by iritkatriel**

### CODE REVIEW

File: Objects/funcobject.c [hunk 1]
```diff
@@ -521,6 +521,9 @@ func_get_doc(PyFunctionObject *op, void *Py_UNUSED(ignored))
 static int
 func_set_doc(PyFunctionObject *op, PyObject *value, void *Py_UNUSED(ignored))
 {
+    if (value == NULL) {
+        value = Py_None;
```
Comment (gvanrossum) [inline Objects/funcobject.c]: Good catch. NULL means "delete" though, not "set to None", so maybe instead of substituting None, we should use XINCREF?

Comment (iritkatriel) [reply Objects/funcobject.c]: Ah yes, do that.

Comment (iritkatriel) [reply Objects/funcobject.c]: I wondered if we're missing a test, but no, the test checks for None, it can't tell the difference because func_get_doc returns None if it's NULL. However, note that now it also hydrates the code object when it's NULL (but not when it's None). Is it possible to del the docstring on an unhydrated object? If it is, then actually setting to None on del would prevent it being hydrated afterwards just to lookup the docstring.

Comment (iritkatriel) [reply Objects/funcobject.c]: There's actually a correctness concern as well - if it was del'd on the function before the code object was hydrated, then we need to know when hydrating that the docstring should not be overwritten by what's in the code object, because the del is more recent.

Comment (gvanrossum) [reply Objects/funcobject.c]: Yeah, we need to eventually figure out how much backwards compatibility is needed here.

My current POV: let's make this act like an instance variable that is backed by a class variable. E.g.
```py
class C:
    attr = "spam"
x = C()
a.attr  # "spam"
a.attr = "eggs"
a.attr  # "eggs"
del a.attr
a.attr  # "spam"
```
(The instance `__dict__` may or may not contain a value that shadows the value in the class `__dict__`.)

The only thing I'd prefer not to do, in the name of backwards compatibility, is raising an error on using `del f.__doc__` when the func_doc field is NULL. In main, you can delete `f.__doc__` all day long and it will always appear `None`. My proposal is that you can still delete it all day long but it will default to the value from the code object.

I presume that deleting docstrings is not a common pattern, so I doubt that anybody cares. We can always ask Raymond later.

Note that there's another interesting observable change. If you never set `f.__doc__` but you assign `f.__code__`, you will see the docstring (if any) of the most recently assigned code object. In main, you would keep seeing the docstring of the initially assigned code object. I claim that the new behavior is better. If you want to keep the original docstring, use `f.__doc__ = f.__doc__`.

Comment (iritkatriel) [reply Objects/funcobject.c]: > My current POV: let's make this act like an instance variable that is backed by a class variable.


Makes sense. If you want it to be None you can set it to None instead if del.

Comment (gvanrossum) [reply Objects/funcobject.c]: > If you want it to be None you can set it to None instead if del.

And that will work in older Python versions too!
